explorer/interpreter/stack_space.h

```diff
@@ -17,39 +17,24 @@ namespace Internal {
 auto IsStackSpaceNearlyExhausted() -> bool;
 
 // Starts a thread to run the function.
-auto RunWithStackSpaceHelper(llvm::function_ref<void()> fn) -> void;
+auto ReserveStackAndRunHelper(llvm::function_ref<void()> fn) -> void;
 
 }  // namespace Internal
 
-// Initializes stack space information for RunWithStackSpace by creating a new
-// thread, and runs the function on the new thread. This only needs to be called
-// once within a given recursion.
-//
-// Usage:
-//   return InitStackSpace<ReturnType>([&]() -> ReturnType {
-//         <function body>
-//       });
-template <typename ReturnType>
-auto InitStackSpace(llvm::function_ref<ReturnType()> fn) -> ReturnType {
-  std::optional<ReturnType> result;
-  Internal::RunWithStackSpaceHelper([&] { result = fn(); });
-  return std::move(*result);
-}
-
 // Runs a function. May start a new thread for the function if too much stack
-// space has been consumed since the last time a thread was spawned (by either
-// InitStackSpace or RunWithStackSpace). Require InitStackSpace be used first
-// within the recursion.
+// space has been consumed since the last time a thread was spawned. Always
+// spawns a new thread when it didn't create the current thread.
```

Comment:
How about something like this instead?

> Runs `fn` after ensuring there is a reasonable amount of space left on the stack for it to run in. This may run `fn` in a separate thread if there is not enough space left on the current stack.

Reply:
Done, with the note about the current thread (because the current thread may have enough space, but with no ability to verify that)